### PR TITLE
Fixes 566 limit body size

### LINKED ISSUE #566: HTTP calls don't have a max size leading to OOM/DoS
@Firstyear prompted me on this one, but if you send a big enough request, it will fill up RAM. This is bad. 😄 

Where testfile.zero is an 8GB file, and the machine has 4GB of RAM, 2GB of swap: 

```
date && time curl -vvvH "Content-Type: application/json" -X POST --data-binary @testfile.zero https://kanidm.example.com/v1/auth && date
<snip>
< HTTP/1.1 100 Continue
* LibreSSL SSL_read: SSL_ERROR_SYSCALL, errno 54
* Closing connection 0
curl: (56) LibreSSL SSL_read: SSL_ERROR_SYSCALL, errno 54
curl -vvvH "Content-Type: application/json" --data-binary    20.56s user 6.29s system 2% cpu 16:07.48 total
```

In the meantime, CPU/Memory usage on the server spiked and the process was eventually OOM terminated by the kernel.

### PR DESCRIPTION
**Pull request by yaleman**
Fixes #566

- [x] cargo fmt has been run
- [x] cargo clippy has been run
- [x] cargo test has been run and passes

### CODE REVIEW

**Comment by yaleman** [comment]
I still want to add testing to this... just have to work out the test harness bit 😆

File: kanidmd/src/lib/core/https/mod.rs [hunk 1]
```diff
@@ -1163,6 +1164,140 @@ impl<State: Clone + Send + Sync + 'static> tide::listener::Listener<State> for T
...
+        req.set_body(body_string);
+
+        let mut res: Response = app.respond(req).await?;
+
+        assert_eq!(tide::StatusCode::BadRequest, res.status());
+        assert_eq!("", res.body_string().await?);
+        Ok(())
+    }
+
+    #[async_std::test]
```
Comment (Firstyear) [inline kanidmd/src/lib/core/https/mod.rs:1216]: You may want to use tokio::test instead here

File: kanidmd/src/lib/core/https/mod.rs [hunk 2]
```diff
@@ -1163,6 +1164,140 @@ impl<State: Clone + Send + Sync + 'static> tide::listener::Listener<State> for T
 }
 */
 
+#[cfg(test)]
+mod tests {
```
Comment (Firstyear) [inline kanidmd/src/lib/core/https/mod.rs:1168]: Tests should be at the bottom of the file.

**Review by Firstyear — CHANGES_REQUESTED** [review]
https://github.com/http-rs/tide/issues/448 It may be better to look at or contribute to this instead of trying our own thing :)

File: kanidmd/src/lib/core/https/mod.rs (hunk 2 above)
Comment (yaleman) [reply kanidmd/src/lib/core/https/mod.rs:1168]: They totally will be once I'm done writing the code - just easier to keep it close and save epic-scrolling on my tiny laptop 😄

File: kanidmd/src/lib/constants/mod.rs [hunk 3]
```diff
@@ -43,3 +43,6 @@ pub const PW_MIN_LENGTH: usize = 10;
 
 // Default
 pub const AUTH_SESSION_EXPIRY: u64 = 3600;
+
+/// Default max request post size, 100MB
+pub const MAX_BODY_SIZE: usize = 104857600;
```
Comment (vDorst) [inline kanidmd/src/lib/constants/mod.rs:48]: I probably unaware why we need a large request size.
100MB seems quit high for this application.
1MB or even 64KB (Max TCP packet size) seems too me, more then enough.

**Comment by yaleman** [comment]
I'm... not even sure why I pushed this as a PR, it's well too early in my own dev cycle for commenting. I'll close it for now, apologies folks 😊

**Comment by Firstyear** [comment]
No apologies needed mate, thanks for looking into it. You should also subscribe to the tide limiting issue :)

File: kanidmd/src/lib/constants/mod.rs (hunk 3 above)
Comment (Firstyear) [reply kanidmd/src/lib/constants/mod.rs:48]: If someone is doing batch changes we'll need more than 1MB in some cases, so it needs to be configurable either per route or per user in some cases.

**Comment by yaleman** [comment]
> No apologies needed mate, thanks for looking into it. You should also subscribe to the tide limiting issue :)

Already have, looks like it's going to take a whole load of coordination amongst projects 😄

**Pull request closed without merging**